client/src/ts/dial.ts

```diff
@@ -0,0 +1,206 @@
+const svgns = 'http://www.w3.org/2000/svg';
+
+// direction to turn the dial in.
+export enum Direction {
+  CounterClockwise = -1,
+  Clockwise = 1
+}
+
+// line ending style.
+export enum LineEnd {
+  Flat = 1,
+  Round = 2,
+}
+
+// animate the curve of the arc using a cubic easing function and animate
+// frames.
+function Animate(start: number, end: number, duration: number, step: number,
+                 callback: (new_val: number) => void): void {
+  let currentIteration = 1;
+  const interations = 60 * duration;
+  if (start > end) {
+    step = step * -1;
+  }
+
+  function easeCubic(pos: number): number {
+    pos /= 0.5;
+    if (pos < 1) {
+      return 0.5 * Math.pow(pos, 3);
+    }
+    return 0.5 * (Math.pow(pos - 2, 3) + 2);
+  }
+
+  function animate(): void {
+    const progress = currentIteration++ / interations;
+    const value = start + step * currentIteration * easeCubic(progress);
+    callback(Math.round(value));
+    if (step > 0 && value < end) {
+      window.requestAnimationFrame(animate);
+    } else if (step < 0 && value > end) {
+      window.requestAnimationFrame(animate);
+    }
+  }
+
+  window.requestAnimationFrame(animate);
+}
+
+// maps polar coordinates to Cartesian to draw the dial.
+function polarToCartesian(centerX: number, centerY: number, radius: number,
+                          angle: number): any {
+  return {
+    x : centerX + radius * Math.cos(angle),
+    y : centerY + radius * Math.sin(angle)
+  };
+}
+
+// describes the SVG arc as a string.
+function describeArc(x: number, y: number, radius: number, startAngle: number,
+                     endAngle: number, reverse: boolean): string {
+  const start = polarToCartesian(x, y, radius, endAngle);
+  const end = polarToCartesian(x, y, radius, startAngle);
+  const largeArcFlag = endAngle - startAngle <= Math.PI ? '0' : '1';
+
+  return [
+    'M', start.x, start.y, 'A', radius, radius, 0, largeArcFlag, 0, end.x, end.y
+  ].join(' ');
+}
+
+// options to adjust the behavior of Dial.
+export class DialOptions {
+  // shape
+  public min: number = 0;
+  public max: number = 100;
+  public step: number = 1;
+  public angleOffset: number = 0;
+  public angleArc: number = 2 * Math.PI;
+  public rotation: Direction = Direction.Clockwise;
+
+  // ui
+  public duration: number = 1;
```

Comment:
Would it make more sense to call this something like animation/transition duration?